pkg/jobworker/worker.go

```diff
@@ -0,0 +1,219 @@
+type Job struct {
+	sync.RWMutex
+	ID      string
+	running bool
+	done    chan bool
+	pgid    int
+	cmd     *exec.Cmd
+	con     ResourceController
+}
+
+// JobOpts wraps the options that can be passed to cgroups for the job
+// details at https://facebookmicrosites.github.io/cgroup2/docs/overview
+type JobOpts struct {
+	CPUWeight int32      // `cpu.weight`
+	IOWeight  int32      // `io.weight`
+	MemLimit  CgroupByte // `mem.high`
+}
+
+// JobStatus is an amalgamation of the useful status information available from the exec.Cmd struct of the job and it's underlying os.Process
+type JobStatus struct {
+	ID       string
+	PID      int
+	Running  bool
+	ExitCode int
+}
+
+func (status JobStatus) String() string {
+	return fmt.Sprintf(`Job Status
+	ID	%s
+	PID	%d
+	Running	%t
+	ExitCode %d`, status.ID, status.PID, status.Running, status.ExitCode)
+}
+
+// ResourceController defines the interface for implementing resource control of new processes
+// In cgroups this will be creating, editing and deleting files in /sys/fs/cgroup
+type ResourceController interface {
+	AddProcess(string, *exec.Cmd) error
+	CreateGroup(string) error
+	DeleteGroup(string) error
+	AddResourceControl(string, JobOpts) error
+}
+
+// NewOpts returns a configured JobOpts based on arguments
+func NewOpts(weight, ioLatency int32, memLimit CgroupByte) JobOpts {
+	return JobOpts{
+		CPUWeight: weight,
+		IOWeight:  ioLatency,
+		MemLimit:  memLimit,
+	}
+}
+
+// NewJob initialises a Job
+func NewJob(id string, cmd *exec.Cmd, con ResourceController) *Job {
+	return &Job{ID: id, running: true, cmd: cmd, con: con, done: make(chan bool, 1)}
+}
+
+// Start calls start using the default ResourceController Cgroup
+func Start(opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	return StartWithController(&Cgroup{"/sys/fs/cgroup"}, opts, cmd, args...)
+}
+
+// StartWithController creates a job's cgroup, adds the resource controls from opts, creates a log file for the cgroup and
+// set's it to the exec.Cmd STDOUT and STDERR. Finally we Start the exec.Cmd and start a go routine to handle the blocking
+// call to Wait so that we can update the job's running flag.
+func StartWithController(con ResourceController, opts JobOpts, cmd string, args ...string) (j *Job, err error) {
+	// Create the job
+	j = NewJob(uuid.New().String(), exec.Command(cmd, args...), con)
+	// Create the cgroup and configure the controllers
+	if err = j.con.CreateGroup(j.ID); err != nil {
+		return nil, fmt.Errorf("failed to create cgroup: %w", err)
+	}
+	// Update cgroup controllers to add resource control to process
+	if err = j.con.AddResourceControl(j.ID, opts); err != nil {
+		return nil, fmt.Errorf("failed to add resource control: %w", err)
+	}
+	// Add job's process to cgroup
+	if err = j.con.AddProcess(j.ID, j.cmd); err != nil {
+		return nil, fmt.Errorf("failed to add PID to cgroup: %w", err)
+	}
+	defer syscall.Close(j.cmd.SysProcAttr.CgroupFD)
```

Comment:
Ah, that answers my question above. I'm assuming that the underlying doesn't actually need that FD longer than `Cmd.Start()` - is that correct?

Reply:
Yes that's correct, once the cmd.Start occurs we can safely close the FD